tidy3d/plugins/adjoint/components/geometry.py

```diff
@@ -0,0 +1,279 @@
+                sim_min_max_val = sim_bounds[min_max_index][dim_index]
+                geo_min_max_val = self.bounds[min_max_index][dim_index]
+                if (min_max_index == 0) and (geo_min_max_val <= sim_min_max_val):
+                    continue
+                if (min_max_index == 1) and (geo_min_max_val >= sim_min_max_val):
+                    continue
+
+                # get the dimensions and edge values on the plane of this surface
+                _, dims_plane = self.pop_axis("xyz", axis=dim_index)
+                _, mins_plane = self.pop_axis(rmin, axis=dim_index)
+                _, maxs_plane = self.pop_axis(rmax, axis=dim_index)
+
+                # construct differential area value and coordinates evenly spaced along this surface
+                d_area = 1.0
+                area_coords = {}
+                for dim_plane, min_edge, max_edge in zip(dims_plane, mins_plane, maxs_plane):
+
+                    # if there is no thickness along this dimension, skip it
+                    length_edge = max_edge - min_edge
+                    if length_edge == 0:
+                        continue
+
+                    # update the differential area value
+                    d_len = length_edge / NUM_CELLS_DIM
+                    d_area *= d_len
+
+                    # construct evenly spaced coordinates along this dimension
+                    interp_vals = np.linspace(
+                        min_edge + d_len / 2, max_edge - d_len / 2, NUM_CELLS_DIM
+                    )
+                    area_coords[dim_plane] = interp_vals
+
+                # for each field component
+                for field_cmp_dim in "xyz":
+
+                    # select the forward and adjoint E fields along this edge
+                    field_name = "E" + field_cmp_dim
+                    e_fwd = grad_data_fwd.field_components[field_name].isel(f=0)
+                    e_adj = grad_data_adj.field_components[field_name].isel(f=0)
+
+                    # select the permittivity data
+                    eps_field_name = f"eps_{field_cmp_dim}{field_cmp_dim}"
+                    eps_data = grad_data_eps.field_components[eps_field_name].isel(f=0)
+
+                    # get the permittivity values just inside and outside the edge
+                    n_cells_in = 2
+                    isel_out = 0 if min_max_index == 0 else -1
+                    isel_ins = n_cells_in if min_max_index == 0 else -n_cells_in - 1
+                    eps2 = eps_data.isel(**{dim_normal: isel_out})
+                    eps1 = eps_data.isel(**{dim_normal: isel_ins})
+
+                    # get gradient contribution for normal component using D field
+                    if field_cmp_dim == dim_normal:
+
+                        # construct normal D fields, dot together at surface
+                        d_fwd = e_fwd * eps_data
+                        d_adj = e_adj * eps_data
+                        d_normal = (d_fwd * d_adj).interp(**normal_coord)
```

Comment:
Just wonder if it's of concern that when the structure is too small, with uniform mesh, there might be just one grid point along one axis in the 2D plane of the gradient monitor so that `interp` will fail because of the single entry issue. One option is to expand the monitor size a bit more in `make_grad_monitors`.

One other suggestion is to explicitly state the interpolation method in `interp` with a constant, in case we want to change to methods other than "linear" in the future.

Reply:
Hm, this is an interesting point, I guess I have not tested with input_structures that are so thin that they would have one grid cell only. (I think this only happens if a size=0)? 

I think how we can handle it is either by erroring if there is one grid cell detected or validating the JaxBox.size to have all non-zero elements. What do you think?

Reply:
For uniform mesh, i wonder if it can happen for size<2*dl. Otherwise, the approach you mentioned is good.

Reply:
I just added a parameter to determine how much to expand the grad monitors as it seemed like the best approach after looking into this. How much do you recommend expanding by? previously it was 1 fp_eps on each side. Do we need at least dL on each side or is this sufficient? 

maybe @momchil-flex knows: basically, what's the minimum size that a FieldMonitor can have before the solver returns only 1 grid cell along that dimension? Does it only need to be > 0 or is this number grid-dependent?

Reply:
Anything larger than 0 will return 2 or even 3 pixels yeah. This is because of the strategy to always have enough data to interpolate to inside the monitor volume.

Reply:
Ok, so even just 2 * fp_eps should be sufficient?  in that case, we should be all set.

Reply:
Yes, the special handling is for `size == 0`.